# Project: Titanic - Machine Learning from Disaster

## I. Import libraries


In [1]:
import pandas as pd
import numpy as np
from IPython.display import display

## II. Load Dataset

In [2]:
titanic_train = pd.read_csv("titanic/train.csv")
display(titanic_train.columns)
titanic_test = pd.read_csv("titanic/test.csv")
display(titanic_test.columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

**Nhận xét:**
- Các cột trong dataset của train và test là giống nhau. Tuy nhiên với train sẽ có cột Outcome là "Survired" để thể hiện số người sống sót (0: mất, 1: sống), còn test thì sẽ không có vì đây là data để thực hiện dự đoán sống sót giữa vào các features.
- Ý nghĩa của từng cột có trong dataset:
<Làm một table cho dễ hiểu>
<Dưới table là một noted để giải thích các cột khcas>


In [3]:
titanic_train.head() #show data in data train

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

**Nhận xét:**
- Trong các cột ở train thì có cột PassengerId đây là cột số thứ tự hành khách tham gia trong chuyến tàu, trong thực tế cột này không có tác dụng trong quá trình build model. 
- Cột này sẽ được thay thế cho vị trí index sẽ hợp lý nhất.

In [4]:
titanic_train.set_index(titanic_train["PassengerId"], inplace= True) # inplace: Trực tiếp trong data
titanic_train.head()

PassengerId  Survived  Pclass  \
PassengerId                                  
1                      1         0       3   
2                      2         1       1   
3                      3         1       3   
4                      4         1       1   
5                      5         0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

**Nhận xét:**
- Khi thực hiện set_index thì cột PassengerId đã trở thành Index, tuy nhiên cột PassengerId vẫn còn trong data, để giải quyết việc này ta sẽ drop cột đó.

In [5]:
titanic_train.drop("PassengerId", axis = 1, inplace= True) # axis = 1: xét theo chiều ngang (đi qua các cột)
titanic_train.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

**Nhận xét:**
- Khi thực hiện drop xong thì ta đã có một dataset về train hoàn thiện về cột. Tuy nhiên cách làm này thực hiện nhiều thao tác, để giải quyết vấn đề này thì ta có cách sau đây

In [6]:
titanic_test = pd.read_csv("titanic/test.csv", index_col= "PassengerId")
titanic_test.head()

Pclass                                          Name     Sex  \
PassengerId                                                                 
892               3                              Kelly, Mr. James    male   
893               3              Wilkes, Mrs. James (Ellen Needs)  female   
894               2                     Myles, Mr. Thomas Francis    male   
895               3                              Wirz, Mr. Albert    male   
896               3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

              Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
PassengerId                                                       
892          34.5      0      0   330911   7.8292   NaN        Q  
893          47.0      1      0   363272   7.0000   NaN        S  
894          62.0      0      0   240276   9.6875   NaN        Q  
895          27.0      0      0   315154   8.6625   NaN        S  
896          22.0      1      1  3101298  12.2875   NaN        S

**Nhận xét:**
- Thực hiện parameter index_col khi đọc dữ liệu vào, và kết quả vẫn giống như dùng cách trên nhưng ngắn gọn hơn. 

## III. Descriptive Statistics (thống kê mô tả)

### (1) **Hiển thị một số thông tin về dữ liệu**

#### 1.1 Số cột và dòng (shape)

In [7]:
titanic_train.shape

(891, 11)

**Nhận xét:**
- Dữ liệu có 891 dòng và gồm 11 cột

#### 1.2 Thông tin dữ liệu

In [8]:
titanic_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Name      891 non-null    object 
 3   Sex       891 non-null    object 
 4   Age       714 non-null    float64
 5   SibSp     891 non-null    int64  
 6   Parch     891 non-null    int64  
 7   Ticket    891 non-null    object 
 8   Fare      891 non-null    float64
 9   Cabin     204 non-null    object 
 10  Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 83.5+ KB


**Nhận xét:**
- Có các cột bị missing values (thiếu dữ liệu) như: Age, Cabin. Các cột này cần kiểm tra lý do mất dữ liệu.

#### 1.3 Kiểu dữ liệu (Dtype)

In [9]:
titanic_train.dtypes

Survived      int64
Pclass        int64
Name         object
Sex          object
Age         float64
SibSp         int64
Parch         int64
Ticket       object
Fare        float64
Cabin        object
Embarked     object
dtype: object

**Nhận xét:**
- Các cột này Dtype (kiểu dữ liệu) chưa chính xác với các cột cần kiểm tra lại và chỉnh sửa (nếu có).

In [10]:
# check unique của các cột trong data train
print("Survived:")
display(titanic_train["Survived"].unique())
print("Pclass:")
display(titanic_train["Pclass"].unique())
print("SibSp:")
display(titanic_train["SibSp"].unique())
print("Parch:")
display(titanic_train["Parch"].unique())
print("Cabin:")
display(titanic_train["Cabin"].unique())
print("Embarked:")
display(titanic_train["Embarked"].unique())

Survived:


array([0, 1], dtype=int64)

Pclass:


array([3, 1, 2], dtype=int64)

SibSp:


array([1, 0, 3, 4, 2, 5, 8], dtype=int64)

Parch:


array([0, 1, 2, 5, 3, 4, 6], dtype=int64)

Cabin:


array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

Embarked:


array(['S', 'C', 'Q', nan], dtype=object)

**Nhận xét:**
- Dựa vào các giá trị có trong các cột trên thì các cột $Survived$, $Pclass$, $SibSp$, $Parch$ và $Embarked$ là các cột có giá trị phân loại (categorical) không phải giá trị số (numerical) hoặc giá trị object cần chuyển đổi kiểu dữ liệu của các cột này.

In [11]:
categorys_train = ["Survived", "Pclass", "SibSp", "Parch", "Embarked"] # Liệt kê các cột cần chuyển
categorys_test = ["Pclass", "SibSp", "Parch", "Embarked"] # Liệt kê các cột cần chuyển, do test không có cột Survived

# define train
def convert_cate_train(df, categorys_train): # tạo hàm có tham số df (dataframe) và categorys là các cột cần chuyển
    for category in categorys_train: # tạo biến catergory và duyệt qua toàn bộ cột có trong categorys 
        df[category] = df[category].astype("category") # các cột cần chuyển sẽ chuyển đổi bằng astype("datatype")
# define test
def convert_cate_test(df, categorys_test): # y như train chỉ khác cột cần chuyển
    for category in categorys_test: 
        df[category] = df[category].astype("category") 

convert_cate_train(titanic_train, categorys_train) # thực hiện gọi hàm và thay thế các parameter
convert_cate_test(titanic_test, categorys_test) 

# show
display(titanic_train.info())
display(titanic_test.info())

<class 'pandas.core.frame.DataFrame'>
Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Survived  891 non-null    category
 1   Pclass    891 non-null    category
 2   Name      891 non-null    object  
 3   Sex       891 non-null    object  
 4   Age       714 non-null    float64 
 5   SibSp     891 non-null    category
 6   Parch     891 non-null    category
 7   Ticket    891 non-null    object  
 8   Fare      891 non-null    float64 
 9   Cabin     204 non-null    object  
 10  Embarked  889 non-null    category
dtypes: category(5), float64(2), object(4)
memory usage: 54.2+ KB


None

<class 'pandas.core.frame.DataFrame'>
Index: 418 entries, 892 to 1309
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Pclass    418 non-null    category
 1   Name      418 non-null    object  
 2   Sex       418 non-null    object  
 3   Age       332 non-null    float64 
 4   SibSp     418 non-null    category
 5   Parch     418 non-null    category
 6   Ticket    418 non-null    object  
 7   Fare      417 non-null    float64 
 8   Cabin     91 non-null     object  
 9   Embarked  418 non-null    category
dtypes: category(4), float64(2), object(4)
memory usage: 25.5+ KB


None

**Nhận xét:**
- Sau khi thực hiện chuyển đổi kiểu dữ liệu của data train và test thì bây giờ kiểu dữ liệu đã đúng cho 2 dataset này.  

#### 1.4 Describe (mô tả thống kê dữ liệu)

1) Numerical data

In [12]:
titanic_train.describe()

Age        Fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

**Nhận xét:**
- Đối với Age: 
    - Độ tuổi nhỏ nhất là 4 tháng tuổi và lớn nhất là 80 tuổi, cùng với đó thì độ tuổi trung bình là 30 tuổi
- Đối với Fare:
    - Dựa vào giá trị min max và mean thì dữ liệu của cột này rời rạc.

2) Categorical data

In [13]:
titanic_train.describe(include=['category'])

Survived  Pclass  SibSp  Parch Embarked
count        891     891    891    891      889
unique         2       3      7      7        3
top            0       3      0      0        S
freq         549     491    608    678      644

**Nhận xét:**
- Đối với Survived: 
    - Bao gồm 0 và 1 đại diện cho mất và sống, tuy nhiên số lượng người mất là cao nhất có tới 549 thiệt mạng
- Đối với Pclass:
    - Hạng vé 3 cao nhất với 491 người, dễ hiểu khi số lượng vé 1 và 2 thường chiếm thiểu số.
- Đối với SibSp:
    - Có tới 608 khách không đi cùng anh chị em hoặc vợ chồng
- Đối với Parch:
    - giống như SibSp có tới 678 các hành khách không đi cùng cha mẹ hoặc con cái của họ.
- Đối với Embarked:
    - Số lượng khách hàng của cảng S chiếm đa số (644)

In [14]:
titanic_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Survived  891 non-null    category
 1   Pclass    891 non-null    category
 2   Name      891 non-null    object  
 3   Sex       891 non-null    object  
 4   Age       714 non-null    float64 
 5   SibSp     891 non-null    category
 6   Parch     891 non-null    category
 7   Ticket    891 non-null    object  
 8   Fare      891 non-null    float64 
 9   Cabin     204 non-null    object  
 10  Embarked  889 non-null    category
dtypes: category(5), float64(2), object(4)
memory usage: 54.2+ KB
